File: android/app/src/main/java/com/fitrater/app/ui/screens/camera/CameraMenuSheet.kt
package com.fitrater.app.ui.screens.camera

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.animation.expandVertically
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.shrinkVertically
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AutoAwesome
import androidx.compose.material.icons.filled.CameraAlt
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.ExpandMore
import androidx.compose.material.icons.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.LocalFireDepartment
import androidx.compose.material.icons.filled.Palette
import androidx.compose.material.icons.filled.People
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.fitrater.app.data.Supa
import com.fitrater.app.ui.theme.HemColors
import com.fitrater.app.ui.theme.HemSpace
import com.fitrater.app.ui.theme.HemType
import com.fitrater.app.ui.theme.PrimaryButton
import com.fitrater.app.ui.theme.SerifDisplay

/** One-tap menu shown when the bottom-nav camera FAB is pressed. */
@Composable
fun CameraMenuSheet(
    onClose: () -> Unit,
    onPickScore: () -> Unit,
    onPickTryOn: () -> Unit,
    onPickVersus: () -> Unit,
    onPickRoast: () -> Unit,
    onPickDecode: () -> Unit,
    onOpenPaywall: () -> Unit = {},
    isPro: Boolean = false,
) {
    // Only one accordion open at a time. null = all collapsed.
    var openId by remember { mutableStateOf<String?>(null) }

    Column(
        Modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = HemSpace.gutter, vertical = HemSpace.md),
    ) {
        Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            Column(Modifier.padding(end = HemSpace.sm)) {
                Text(
                    "TONIGHT'S TOOLS",
                    style = HemType.smallLabel.copy(color = HemColors.Bronze, letterSpacing = 2.sp),
                )
                Spacer(Modifier.height(HemSpace.xs))
                SerifDisplay("What's the look for?")
            }
            Spacer(Modifier.fillMaxWidth().weight(1f, fill = false))
            Box(
                Modifier.size(36.dp).clickable(onClick = onClose),
                contentAlignment = Alignment.Center,
            ) { Icon(Icons.Default.Close, contentDescription = "Close") }
        }
        Spacer(Modifier.height(HemSpace.lg))

        val toggle: (String) -> Unit = { id -> openId = if (openId == id) null else id }

        MenuHairline()
        AccordionRow(
            id = "score",
            icon = Icons.Filled.CameraAlt,
            title = "Score a look",
            subtitle = "One photo, one honest number.",
            cost = "${Supa.SCORE_COST} credits",
            description = "Snap or upload a fit and Hem calls it — score, per-piece notes, and one line of verdict.",
            needs = "1 photo of the outfit.",
            open = openId == "score",
            onTap = { toggle("score") },
            onGo = { onPickScore() },
        )
        MenuHairline()
        AccordionRow(
            id = "tryon",
            icon = Icons.Filled.AutoAwesome,
            title = "Try on",
            subtitle = "Wear a Studio piece on your own photo.",
            cost = "${Supa.TRYON_COST} credits",
            description = "Wear a Studio piece — or a garment you upload — on your own photo. Hem does the compositing.",
            needs = "1 photo of yourself + 1 garment.",
            open = openId == "tryon",
            onTap = { toggle("tryon") },
            onGo = { if (isPro) onPickTryOn() else onOpenPaywall() },
            goLabel = if (isPro) "GO →" else "Start 7-day trial →",
            proLocked = !isPro,
        )
        MenuHairline()
        AccordionRow(
            id = "versus",
            icon = Icons.Filled.People,
            title = "A vs B",
            subtitle = "Two photos, one winner.",
            cost = "${Supa.VERSUS_COST} credits",
            description = "Two fits, one winner. Hem picks and explains, sharp and one-sentence.",
            needs = "2 photos.",
            open = openId == "versus",
            onTap = { toggle("versus") },
            onGo = { onPickVersus() },
        )
        MenuHairline()
        AccordionRow(
            id = "roast",
            icon = Icons.Filled.LocalFireDepartment,
            title = "Roast this",
            subtitle = "Brutal mode, shareable card.",
            cost = "${Supa.ROAST_COST} credits",
            description = "Brutal mode, locked. Hem roasts the outfit, never the person, and always ends with a fix.",
            needs = "1 photo.",
            open = openId == "roast",
            onTap = { toggle("roast") },
            onGo = { onPickRoast() },
        )
        MenuHairline()
        AccordionRow(
            id = "decode",
            icon = Icons.Filled.Palette,
            title = "Decode style",
            subtitle = "Any reference — get the recipe.",
            cost = "${Supa.DECODE_COST} credits",
            description = "Any reference — celebrity, magazine, IG. Hem breaks down the pieces, palette, and one-line style signature.",
            needs = "1 reference photo.",
            open = openId == "decode",
            onTap = { toggle("decode") },
            onGo = { onPickDecode() },
        )
        MenuHairline()
        Spacer(Modifier.height(HemSpace.xl))
    }
}

@Composable
private fun MenuHairline() {
    Box(Modifier.fillMaxWidth().height(1.dp).background(HemColors.Hairline))
}

@Composable
private fun AccordionRow(
    id: String,
    icon: ImageVector,
    title: String,
    subtitle: String,
    cost: String,
    description: String,
    needs: String,
    open: Boolean,
    onTap: () -> Unit,
    onGo: () -> Unit,
    goLabel: String = "GO →",
    proLocked: Boolean = false,
) {
    Column(Modifier.fillMaxWidth()) {
        Row(
            Modifier
                .fillMaxWidth()
                .clickable(onClick = onTap)
                .padding(vertical = HemSpace.md),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                icon,
                contentDescription = null,
                tint = HemColors.Ink,
                modifier = Modifier.size(24.dp),
            )
            Spacer(Modifier.size(HemSpace.md))
            Column(Modifier.weight(1f)) {
                Text(title, style = HemType.serifSection.copy(fontSize = 20.sp))
                Spacer(Modifier.height(2.dp))
                Text(subtitle, style = HemType.bodyMuted.copy(fontSize = 13.sp))
            }
            Spacer(Modifier.size(HemSpace.sm))
            Column(horizontalAlignment = Alignment.End) {
                Text(
                    cost.uppercase(),
                    style = HemType.smallLabel.copy(color = HemColors.Bronze, letterSpacing = 1.5.sp),
                )
                if (proLocked) {
                    Spacer(Modifier.height(4.dp))
                    Box(
                        Modifier
                            .clip(RoundedCornerShape(999.dp))
                            .background(HemColors.Bronze)
                            .padding(horizontal = 6.dp, vertical = 2.dp),
                    ) {
                        Text(
                            "PRO",
                            style = HemType.smallLabel.copy(
                                color = Color.White,
                                letterSpacing = 1.5.sp,
                                fontSize = 10.sp,
                            ),
                        )
                    }
                }
            }
            Spacer(Modifier.size(HemSpace.xs))
            // Chevron: right when collapsed, down when open. Rotated for a smooth transition.
            val icn = if (open) Icons.Default.ExpandMore else Icons.Default.KeyboardArrowRight
            Icon(
                icn,
                contentDescription = null,
                tint = HemColors.Bronze,
                modifier = Modifier.size(22.dp).rotate(0f),
            )
        }
        AnimatedVisibility(
            visible = open,
            enter = fadeIn(tween(180)) + expandVertically(
                animationSpec = spring(dampingRatio = 0.8f, stiffness = Spring.StiffnessMediumLow),
            ),
            exit = fadeOut(tween(120)) + shrinkVertically(tween(180)),
        ) {
            Column(
                Modifier
                    .fillMaxWidth()
                    .padding(start = HemSpace.md + 24.dp, end = HemSpace.xs, bottom = HemSpace.md),
            ) {
                Text(
                    description,
                    style = HemType.body.copy(fontSize = 14.sp, lineHeight = 20.sp),
                )
                Spacer(Modifier.height(HemSpace.sm))
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        "NEEDS",
                        style = HemType.smallLabel.copy(color = HemColors.Bronze, letterSpacing = 1.5.sp),
                    )
                    Spacer(Modifier.size(HemSpace.xs))
                    Text(
                        needs,
                        style = HemType.bodyMuted.copy(fontSize = 13.sp, fontWeight = FontWeight.Medium),
                    )
                }
                Spacer(Modifier.height(HemSpace.md))
                Row(Modifier.fillMaxWidth(), horizontalArrangement = androidx.compose.foundation.layout.Arrangement.End) {
                    Box(Modifier.fillMaxWidth(if (proLocked) 0.75f else 0.55f)) {
                        PrimaryButton(label = goLabel, onClick = onGo)
                    }
                }
            }
        }
    }
}

File: ios/FitScore/FitScore/Camera/CameraMenuSheet.swift
import SwiftUI

/// Accordion sheet — one tile per Camera flow (Score / Try-on / A vs B / Roast
/// / Decode). Only one tile expanded at a time. Ports Android
/// `CameraMenuSheet` behavior + copy.
struct CameraMenuSheet: View {
    @Environment(\.dismiss) private var dismiss
    @EnvironmentObject var session: SessionStore

    @State private var openId: String?

    @State private var showScore = false
    @State private var showTryOn = false
    @State private var showVersus = false
    @State private var showDecode = false
    @State private var showPaywall = false
    @State private var paywallContext: PaywallContext? = nil
    @State private var occasionFreeAvailable = false

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                HStack(alignment: .center) {
                    VStack(alignment: .leading, spacing: 6) {
                        Text("TONIGHT'S TOOLS")
                            .font(.system(size: 10, weight: .semibold))
                            .tracking(2)
                            .foregroundStyle(Palette.bronze)
                        Text("What's the look for?")
                            .font(Serif.display(26))
                            .foregroundStyle(Palette.ink)
                    }
                    Spacer()
                    Button(action: { dismiss() }) {
                        Image(systemName: "xmark")
                            .foregroundStyle(Palette.ink)
                            .frame(width: 36, height: 36)
                    }
                }
                Spacer(minLength: 20)

                Hairline()
                row("score", icon: "camera.fill",
                    title: "Score a look",
                    subtitle: "One photo, one honest number.",
                    cost: "\(Supa.scoreCost) credits",
                    description: "Snap or upload a fit and Hem calls it — score, per-piece notes, and one line of verdict.",
                    needs: "1 photo of the outfit.",
                    goLabel: "GO →",
                    proLocked: false,
                    onGo: {
                        dismiss()
                        DispatchQueue.main.asyncAfter(deadline: .now() + 0.35) { CameraMenuBus.shared.request(.score) }
                    })
                Hairline()
                row("tryon", icon: "sparkles",
                    title: "Try on",
                    subtitle: "Wear a Studio piece on your own photo.",
                    cost: "\(Supa.tryonCost) credits",
                    description: "Wear a Studio piece — or a garment you upload — on your own photo. Hem does the compositing.",
                    needs: "1 photo of yourself + 1 garment.",
                    goLabel: RcBilling.shared.isPro ? "GO →" : "Start 7-day trial →",
                    proLocked: !RcBilling.shared.isPro,
                    onGo: {
                        let ctx = FeatureGates.requireTryon()
                        dismiss()
                        DispatchQueue.main.asyncAfter(deadline: .now() + 0.35) {
                            if let ctx { CameraMenuBus.shared.request(.paywall, context: ctx) }
                            else { CameraMenuBus.shared.request(.tryon) }
                        }
                    })
                Hairline()
                row("versus", icon: "person.2.fill",
                    title: "A vs B",
                    subtitle: "Two photos, one winner.",
                    cost: "\(Supa.versusCost) credits",
                    description: "Two fits, one winner. Hem picks and explains, sharp and one-sentence.",
                    needs: "2 photos.",
                    goLabel: "GO →",
                    proLocked: false,
                    onGo: {
                        dismiss()
                        DispatchQueue.main.asyncAfter(deadline: .now() + 0.35) { CameraMenuBus.shared.request(.versus) }
                    })
                Hairline()
                row("chat", icon: "bubble.left.and.bubble.right.fill",
                    title: "Chat with Hem",
                    subtitle: "Ask anything — fit, occasion, swap ideas.",
                    cost: "free",
                    description: "Your stylist on tap. Send a photo or a question — Hem answers honestly, in one voice. Remembers your recent fits.",
                    needs: "Nothing. Just start typing.",
                    goLabel: "OPEN →",
                    proLocked: false,
                    onGo: {
                        dismiss()
                        DispatchQueue.main.asyncAfter(deadline: .now() + 0.35) { CameraMenuBus.shared.request(.chat) }
                    })
                Hairline()
                row("decode", icon: "paintpalette.fill",
                    title: "Decode style",
                    subtitle: "Any reference — get the recipe.",
                    cost: "\(Supa.decodeCost) credits",
                    description: "Any reference — celebrity, magazine, IG. Hem breaks down the pieces, palette, and one-line style signature.",
                    needs: "1 reference photo.",
                    goLabel: "GO →",
                    proLocked: false,
                    onGo: {
                        dismiss()
                        DispatchQueue.main.asyncAfter(deadline: .now() + 0.35) { CameraMenuBus.shared.request(.decode) }
                    })
                Hairline()
                row("occasion", icon: "sparkle.magnifyingglass",
                    title: "Occasion Coach",
                    subtitle: "Big day? Hem plans 3 fits.",
                    cost: occasionFreeAvailable ? "free this week" : "\(Supa.occasionCost) credits",
                    description: "Tell Hem the moment. He picks 3 combos from your closet and gaps for Studio to fill.",
                    needs: "1 sentence about the occasion.",
                    goLabel: "GO →",
                    proLocked: false,
                    onGo: {
                        dismiss()
                        DispatchQueue.main.asyncAfter(deadline: .now() + 0.35) { CameraMenuBus.shared.request(.occasion) }
                    })
                // Covers are shelved, so the whole accordion entry — including its
                // leading hairline — drops out to avoid a doubled divider here.
                if Supa.magazineCoversEnabled {
                    Hairline()
                    row("cover", icon: "text.book.closed.fill",
                        title: "Compose a cover",
                        subtitle: "Your fit, on the cover.",
                        cost: "\(Supa.magazineCoverCost) credits",
                        description: "Turn any fit into an editorial magazine cover — masthead, headline, pull-quote. Made to be shared.",
                        needs: "1 photo (or pick from Journal).",
                        goLabel: "GO →",
                        proLocked: false,
                        onGo: {
                            dismiss()
                            DispatchQueue.main.asyncAfter(deadline: .now() + 0.35) { CameraMenuBus.shared.request(.cover) }
                        })
                }
                Hairline()
                row("invitation", icon: "envelope.open.fill",
                    title: "Decode the invite",
                    subtitle: "Dress code → 3 closet combos.",
                    cost: "\(Supa.invitationDecodeCost) credits",
                    description: "Snap the invitation. Hem reads the dress code and pulls 3 combinations from your Studio closet.",
                    needs: "1 photo of the invitation.",
                    goLabel: "GO →",
                    proLocked: false,
                    onGo: {
                        dismiss()
                        DispatchQueue.main.asyncAfter(deadline: .now() + 0.35) { CameraMenuBus.shared.request(.invitation) }
                    })
                Hairline()
                Spacer(minLength: 30)
            }
            .padding(.horizontal, 20)
            .padding(.top, 16)
        }
        .background(Palette.paper.ignoresSafeArea())
        .task {
            occasionFreeAvailable = await OccasionCoachService.weeklyFreeAvailable()
        }
    }

    @ViewBuilder
    private func row(_ id: String,
                     icon: String,
                     title: String,
                     subtitle: String,
                     cost: String,
                     description: String,
                     needs: String,
                     goLabel: String,
                     proLocked: Bool,
                     onGo: @escaping () -> Void) -> some View {
        let open = openId == id
        VStack(alignment: .leading, spacing: 0) {
            Button(action: {
                Haptic.chip()
                withAnimation(.spring(response: 0.35, dampingFraction: 0.85)) {
                    openId = open ? nil : id
                }
            }) {
                HStack(alignment: .center, spacing: 12) {
                    Image(systemName: icon)
                        .font(.system(size: 20))
                        .foregroundStyle(Palette.ink)
                        .frame(width: 24, height: 24)
                    VStack(alignment: .leading, spacing: 3) {
                        Text(title)
                            .font(Serif.display(20))
                            .foregroundStyle(Palette.ink)
                        Text(subtitle)
                            .font(Serif.body(13))
                            .foregroundStyle(Palette.muted)
                    }
                    Spacer()
                    VStack(alignment: .trailing, spacing: 4) {
                        Text(cost.uppercased())
                            .font(.system(size: 10, weight: .semibold))
                            .tracking(1.5)
                            .foregroundStyle(Palette.bronze)
                        if proLocked {
                            Text("PRO")
                                .font(.system(size: 10, weight: .semibold))
                                .tracking(1.5)
                                .foregroundStyle(.white)
                                .padding(.horizontal, 6)
                                .padding(.vertical, 2)
                                .background(Palette.bronze)
                                .clipShape(Capsule())
                        }
                    }
                    Image(systemName: open ? "chevron.down" : "chevron.right")
                        .font(.system(size: 14, weight: .semibold))
                        .foregroundStyle(Palette.bronze)
                        .frame(width: 22, height: 22)
                }
                .padding(.vertical, 14)
                .contentShape(Rectangle())
            }
            if open {
                VStack(alignment: .leading, spacing: 12) {
                    Text(description)
                        .font(Serif.body(14))
                        .foregroundStyle(Palette.ink)
                    HStack(spacing: 8) {
                        Text("NEEDS")
                            .font(.system(size: 10, weight: .semibold))
                            .tracking(1.5)
                            .foregroundStyle(Palette.bronze)
                        Text(needs)
                            .font(Serif.body(13, weight: .medium))
                            .foregroundStyle(Palette.muted)
                    }
                    HStack {
                        Spacer()
                        PrimaryButton(title: goLabel, action: {
                            Haptic.tap()
                            onGo()
                        })
                        .frame(maxWidth: proLocked ? .infinity : 260)
                    }
                }
                .padding(.leading, 36)
                .padding(.trailing, 4)
                .padding(.bottom, 14)
                .transition(.opacity.combined(with: .move(edge: .top)))
            }
        }
    }
}
